nix/default.nix

```diff
@@ -2,17 +2,26 @@
 let
   nix-pre-commit-hooks = (import (builtins.fetchTarball "https://github.com/cachix/pre-commit-hooks.nix/tarball/master/" + "/nix/") { sources = sources; }).packages;
   overlay = _self: pkgs:
-    let sharedOverrides = {
-        overrides = _self: super: {
-            mkDerivation = args: super.mkDerivation (args //
-                {
-                    # skip running tests for Hackage packages
-                    doCheck = args.pname == "ghcide" || args.pname == "haskell-language-server";
-                    # relax upper bounds
-                    jailbreak = args.pname != "jailbreak-cabal";
-                });
-            };
```

Comment:
This was by design, yes. That's why having a working binary is so important. Build times only became slow when the Nixpkgs version was bumped and the Cachix snapshot needed to be regenerated.

Maintaining a list of jailbroken packages is a maintenance overhead. This Nix script has required zero maintenance which in my book is a huge success! Please keep it that way

Reply:
> Maintaining a list of jailbroken packages is a maintenance overhead.

If this is the critical reason why we override `mkDerivation`, I'm not convinced. We are rebuilding massive derivations which we shouldn't have done. For an arbitrary derivation produced from cabal2nix, even if the output from haskell world is exactly identical, nix still considers it as an unknown drv which needs to be built if we override `mkDerivation`. I see our building steps start from `primitive` or whatever really low-level packages in haskell ecosystem, which indicates that we are rebuilding *everything*. This kind of rebuilds are totally meaningless: not only do they take up a lot of CI resource (yeah, thanks to cachix, it only happens when we bump nixpkgs), but also eat a large amount of disk space. I'm worry about the later one. I believe the principle of nix is about providing a nice sandbox mechanism and keeping the storage occupation as small as possible by sharing common drvs between projects. The slogan of cachix is "Never build software more than once." -- wouldn't it be better that we don't need to build at all? The fact is that haskell packages for hls development can be used only right here, though they could be used elsewhere from the haskell sight, whereas nix is preventing us because of mismatched hashes. I cleared my cabal store then tested with this PR by running `nix-shell -- run "cabal update && cabal build all"`,  and then listed packages nix didn't provide: 

```
$ ls ~/.cabal/store/ghc-8.10.4
base16-bytestring-1.0.1.0-d0926e90c2815989331387be96d8f416028ba3bc9a123d94b1d0f274a3a6a5f4
binary-instances-1.0.1-6ba1c718b7533f4c9915c670662a76763f286a1114ec80bad71eb5eed897f173
Cabal-3.4.0.0-2741f2a79654fd73ccdd20cc936d14e79be724f4e118a011afaf2e0251fcbd42
cabal-install-parsers-0.4.1-80d7818a92c05397a7a67ab9c821a4bea0c6f65b16a121c45d9413d59de9646c
cryptohash-sha256-0.11.102.0-e3c853fa4ea0102e499fbc43891eb19b41a4ac25e5dec8777e34a69c6472cf05
ghcide-1.1.0.0-3f904f36ff477a4e22e0fa0d2440e7d02d8bee8d161897ed003465013db3e8a5
haskell-language-server-1.0.0.0-b0de54f048a8eee8efe90bd41563e9206efd01e14a962688712544a76cdc4023
haskell-language-server-1.0.0.0-e-haskell-language-server-d3dd253a06568cfa4b23a608a3ce92055b4d65b056da225d8c004493036c973b
haskell-language-server-1.0.0.0-e-haskell-language-server-wrapper-1abd864224972cf08b8640a4c0e43eeab8639794108138c6676b61243cadc0ca
hie-compat-0.1.0.0-d56830e842acce4e8bac6c7034aa16fb281b853a8c24e15347b4d7f53d301f6d
hiedb-0.3.0.1-ac8a7f6f8ae5427fedd5aba3640c5f56b7a23e83882f2c493364431fdbcf3b76
hls-brittany-plugin-1.0.0.0-7b360aea0b328be60c94b446aae4f8415f3e2851e24857169d74523fa8a1d448
hls-class-plugin-1.0.0.0-a2d168c2ed9e81bf9153531f4380cab844f96cd397b353716e4d0f86ffe0d369
hls-eval-plugin-1.0.0.0-9548052d8ada0295d0807b21a85729237a6f7a2a859fa1f0a05b58666f2cda12
hls-explicit-imports-plugin-1.0.0.0-006beccc87d9d9c37cd098857a0ae6c4a46b0224179fa3394781466b889cad48
hls-haddock-comments-plugin-1.0.0.0-ede16844f1f2a8e95f73e78a2ac2c0c11dbfe3918d18494966962af53f2b46ec
hls-hlint-plugin-1.0.0.1-115c12a35a0041a6f340a8e13bcc9bc1c4b310d0c26c49288e0a45c7c743e4ee
hls-plugin-api-1.0.0.0-60c4897af6be2c5abec46d3e7f29e010797fb29a6e56df0b8a240fde75197db7
hls-retrie-plugin-1.0.0.0-3b007d77291ec6a90fa5b67862a64bed61e0035b6955df25a1dff07e73edfee4
hls-splice-plugin-1.0.0.0-78e9daa30906ee11a808c497cef74b0186102ea51f72b8714f9d997730cb90e9
hls-tactics-plugin-1.0.0.0-e44eb44963f274f795e1b04a6188b96e2fffac51d6b8accb2c6d908287538690
hsc2hs-0.68.7-e-hsc2hs-a54e898f36feb5209908d7942eb1f2ba686bb25c99af0ac9b04c8f9f55182b1e
hspec-discover-2.7.8-77777f053cfe070a34b1bd8bb9b8037f6f4551c9d413c30af38b85ccb93abea0
hspec-discover-2.7.8-e-hspec-discover-44ebe0b621b3c877a55a62dc9acf89c29c1338ac4d29f079f533ccec23f1a3ac
implicit-hie-0.1.2.5-b4349caf1b850fcd8a6d44c0865cece28474013c5e2d94dbaec0d8346846ea2b
implicit-hie-0.1.2.5-e-gen-hie-7d763293798489a88ca990b1f811c0bfefeb3a6e5c3eca24dc1f971ffe2eda65
incoming
lsp-test-0.13.0.0-251e4ec678698ba1c82e379ff42d22939692158816afe9f39bdc5c6525e7f1e3
lukko-0.1.1.3-9f8e3d93d3cda8c84063d31c8e069269620a0160c76e8d5b0fb6c4eac8e4c061
package.db
shake-0.18.5-2d226d4377c6371fb193afa7fe061a74140921e7eaa2b2ff578bdd1b8d06eb82
tar-0.5.1.1-18946a119ecc2a1fbcf68d5238b97fbf3bba2eafaf966c3c485e77eb1e2830cc
text-binary-0.2.1.1-774ea2d642d52009c735fb73cf6c601eae465c22b6abb821d7203b2da7bdb6bc
vector-binary-instances-0.2.5.1-80b725e9b74a9ea805f61221fc88cc6c7fd99ade550d2c852b6aa471486c1efd
```

Except packages we were developing (like hls itself or its plugins), only few packages whose versions in nixpkgs were not in range were left to `cabal-install`. Compared to the old situation, some packages which broke the consistency of cabal store (see https://gitlab.haskell.org/ghc/ghc/-/issues/16996#note_330276) had gone that I was not able to reproduce #1365.  As for jailbroken packages, ideally they should be revised in upstream. Currently all jailbroken packages are introduced by `shake-bench`, however, many of theirs upstreams are not seem to be responsive. Thus we could send PR to nixpkgs to jailbreak, or patch cabal files -- which wouldn't be hard -- if we don't want to do this in hls.

Reply:
cc @maralorn, I remember you are one of hls maintainers of nixpkgs

Reply:
I agree with both of you!

1. That global override solution is terribly inefficient and frustrating to me. This should not be necessary and increases my dissatisfaction with the current state of nixpkgs.

2. If the list of overrides that you are adding below is actually needed I totally agree with pepe that this is a maintenance overhead, and not a really small one. So I totally understand where this override comes from.

But I must say I am surprised it’s this bad. We always have hls building in nixpkgs. So all of it’s dependencies should be in there. The only problem I can think of is when you pick up a new dependency which is at the moment unmaintained and therefore likely broken in nixpkgs. That is e.g. true for hiedb right now. (It’s already fixed but not yet released to the unstable channel.) But there are so many more overrides there, it’s confusing to me. Is this derivation building anything more than hls and it's dependencies?

If you have any suggestions how we could improve that situation from upstream nixpkgs I‘d be all ears.

Reply:
> But there are so many more overrides there, it’s confusing to me. Is this derivation building anything more than hls and it's dependencies?

Except `hiedb`, packages I override are brought by  `shake-bench`,  a dependency of  benchmark component `benchHist`. But we don't `doBenchmark` by default for haskell drvs, right? So this problem arises in developing, not distributing.

Reply:
Ah, yeah. It would be so great if shake-bench would be released to hackage. Then it would be a dependency of ghcide and I would gladly fix all it's dependencies.

Reply:
And at least we could unbreak these diagram/svg packages in nixpkgs

Reply:
`shake-bench` is in Hackage now